WIP: Fix icolamp in export_cpp.py (against gpucpp)

valassi opened the pull request:
Hi @oliviermattelaer again I am reposting my previous #115, as you modified it to be against gpucpp_june24 and to include stuff now in gpucpp_june24.

This new MR *is meant to be against gpucpp*, not gpucpp_june24. I keep #115 open for later discussions about gpucpp_june24.

Can you please review (with respect to gpucpp)?

Thanks
Andrea

valassi committed:
In export_cpp.py fix bug #114 in get_icolamp_lines, resulting in different icolamp arrays for F77 and CPP (see madgraph5/madgraph4gpu#873)

valassi committed:
minor fix in a printout in my previous patch in export_cpp.py

valassi commented:
Hi @oliviermattelaer I have accepted your suggestions in https://github.com/madgraph5/madgraph4gpu/pull/877 (thanks!), with small extra fine tunings (that go in the same direction you suggested), so I will soon be able to merge it.

Can you review this one please before, https://github.com/mg5amcnlo/mg5amcnlo/pull/116 ?

In https://github.com/madgraph5/madgraph4gpu/pull/877 I am referencing the gpucpp code at this commit, so it would be best to merge this first. It is a bug fix, I hope it is not controversial? Thanks Andrea

oliviermattelaer commented on madgraph/iolibs/export_cpp.py:
```diff
@@ -1770,6 +1770,7 @@ def get_icolamp_lines(self, mapconfigs, matrix_element, num_matrix_element):
         booldict = {False: "false", True: "true"}
 
         # Only want to include leading color flows, so find max_Nc
+        misc.sprint('get_icolamp_lines CPP',mapconfigs)
```
```suggestion
```

oliviermattelaer commented on madgraph/iolibs/export_cpp.py:
```diff
@@ -1802,11 +1803,16 @@ def get_icolamp_lines(self, mapconfigs, matrix_element, num_matrix_element):
                                           [ijamp+1]
                 #else:
                 #    self.proc_characteristic['single_color'] = False
+        ###misc.sprint('get_icolamp_lines CPP',diag_jamp)
```
```suggestion
```

oliviermattelaer commented on madgraph/iolibs/export_cpp.py:
```diff
@@ -1802,11 +1803,16 @@ def get_icolamp_lines(self, mapconfigs, matrix_element, num_matrix_element):
+        for iconfig, num_diag_key in enumerate(mapconfigs):  
+            # [AV fix bug #114: icolamp array is different in Fortran and CPP]
+            # Variable mapconfigs can be a list or a dictionary
+            # In the old (buggy) implementation num_diag was the key of the dictionary
```
This line is useless since the old code is not kept...
```suggestion
```

oliviermattelaer commented on madgraph/iolibs/export_cpp.py:
```diff
@@ -1802,11 +1803,16 @@ def get_icolamp_lines(self, mapconfigs, matrix_element, num_matrix_element):
+            # Variable mapconfigs can be a list or a dictionary
+            # In the old (buggy) implementation num_diag was the key of the dictionary
+            # In the new (fixed) implementation num_diag is the value of the dictionary, extracting the first list element, and adding +1 (F77/CPP indexing?)
+            num_diag = mapconfigs[num_diag_key][0]+1
```
Does this work if mapconfigs is a list?
I would need to investigate this...

oliviermattelaer commented on madgraph/iolibs/export_cpp.py:
```diff
@@ -1802,11 +1803,16 @@ def get_icolamp_lines(self, mapconfigs, matrix_element, num_matrix_element):
+            # In the old (buggy) implementation num_diag was the key of the dictionary
+            # In the new (fixed) implementation num_diag is the value of the dictionary, extracting the first list element, and adding +1 (F77/CPP indexing?)
+            num_diag = mapconfigs[num_diag_key][0]+1
+            ###misc.sprint('get_icolamp_lines CPP',iconfig,num_diag)
```
```suggestion
```

oliviermattelaer commented on madgraph/iolibs/export_v4.py:
```diff
@@ -1186,6 +1186,7 @@ def get_icolamp_lines(self, mapconfigs, matrix_element, num_matrix_element):
 
         booldict = {False: ".false.", True: ".true."}
 
+        misc.sprint('get_icolamp_lines F77',mapconfigs)
```
```suggestion
```

oliviermattelaer commented on madgraph/iolibs/export_v4.py:
```diff
@@ -1224,9 +1225,11 @@ def get_icolamp_lines(self, mapconfigs, matrix_element, num_matrix_element):
                                           [ijamp+1]
                 else:
                     self.proc_characteristic['single_color'] = False
+        ###misc.sprint('get_icolamp_lines F77',diag_jamp)
```
```suggestion
```

oliviermattelaer commented on madgraph/iolibs/export_v4.py:
```diff
@@ -1224,9 +1225,11 @@ def get_icolamp_lines(self, mapconfigs, matrix_element, num_matrix_element):
 
         colamps = ijamp + 1
         for iconfig, num_diag in enumerate(mapconfigs):        
+            ###misc.sprint('get_icolamp_lines F77',iconfig,num_diag)
```
```suggestion
```

oliviermattelaer reviewed (changes requested):
Hi Andrea, 

This is not a trivial PR, that can be review like this.

From the comment of the code, it sounds that you might not handle all the case correctly (likely you handle group mode gpucpp style correctly but not the simpler non group mode or are not ready here for the full group mode --that we do not support yet in gpucpp).

I have put the "cleaning" stuff that are needed and point where the change is not clear to me and where I need to investigate it.
In any case, this is a quite independent PR compare to the main PR that we have to handle right now (since this only affect the colour writing of some event --if not then I do not understand this PR and therefore I can not approve it anyway). 
So I would put this in low priority for the moment and focus to allow the merging/PR of gpucpp_june24 within 3.6.0 and focus on the mismatch of cross-section (with Stefan) which I think are more urgent for the moment.

Sorry to stall on this. I do believe that you probably cover all the cases (because by default I do trust you) but the point of those PR is to double check that and to carefully advance one step at the time.

valassi commented:
> So I would put this in low priority for the moment and focus to allow the merging/PR of gpucpp_june24 within 3.6.0 and focus on the mismatch of cross-section (with Stefan) which I think are more urgent for the moment.

I totally disagree. I could not disagree more. I spent the whole of last week to fix #856 and it is done. All tests in cudacpp pass in the CI. I do not intend to go back on this. I would just have totally wasted more than a week of hard work.

If you do not want this in the mg5amcnlo, I suggest that I put this in cudacpp, as I have done in the past.

valassi commented:
> Hi Andrea,
> 
> This is not a trivial PR, that can be review like this.
> 
> From the comment of the code, it sounds that you might not handle all the case correctly (likely you handle group mode gpucpp style correctly but not the simpler non group mode or are not ready here for the full group mode --that we do not support yet in gpucpp).
> 
> I have put the "cleaning" stuff that are needed and point where the change is not clear to me and where I need to investigate it. In any case, this is a quite independent PR compare to the main PR that we have to handle right now (since this only affect the colour writing of some event --if not then I do not understand this PR and therefore I can not approve it anyway). So I would put this in low priority for the moment and focus to allow the merging/PR of gpucpp_june24 within 3.6.0 and focus on the mismatch of cross-section (with Stefan) which I think are more urgent for the moment.
> 
> Sorry to stall on this. I do believe that you probably cover all the cases (because by default I do trust you) but the point of those PR is to double check that and to carefully advance one step at the time.

Hi Olivier, thanks for looking at this.

I disagree with some things you said, especially the priority. I spent my time fixing #856, all tests pass, I do not want to go back. It feels like wasting ten days of hard work, I cannot do that.

I understand you do not like to put this in mg5amcnlo yet however. I suggest then that I will recreate the patch in cudacpp (as I did for so many other parts of export_cpp that I overload there). I will modify https://github.com/madgraph5/madgraph4gpu/pull/877 accordingly. Then the backport from cudacpp to mg5amcnlo can be done when you prefer.

oliviermattelaer commented:
The problem is not where you do the patch here, but that I need a careful check of such PR
which is highly non trivial and can not be approve like that either part of the plugin or of mg5amc....

So I can/will do it but not now tonight or super high priority. This will have to wait a bit.

Olivier




On 2 Jul 2024, at 22:14, Andrea Valassi ***@***.***> wrote:



Hi Andrea,

This is not a trivial PR, that can be review like this.

From the comment of the code, it sounds that you might not handle all the case correctly (likely you handle group mode gpucpp style correctly but not the simpler non group mode or are not ready here for the full group mode --that we do not support yet in gpucpp).

I have put the "cleaning" stuff that are needed and point where the change is not clear to me and where I need to investigate it. In any case, this is a quite independent PR compare to the main PR that we have to handle right now (since this only affect the colour writing of some event --if not then I do not understand this PR and therefore I can not approve it anyway). So I would put this in low priority for the moment and focus to allow the merging/PR of gpucpp_june24 within 3.6.0 and focus on the mismatch of cross-section (with Stefan) which I think are more urgent for the moment.

Sorry to stall on this. I do believe that you probably cover all the cases (because by default I do trust you) but the point of those PR is to double check that and to carefully advance one step at the time.

Hi Olivier, thanks for looking at this.

I disagree with many things you said, especially the priority. I spent my time fixing #856, all tests pass, I do not want to go back. It feels like wasting ten days of hard work, I cannot do that.

I understand you do not like to put this in mg5amcnlo yet however. I suggest then that I will recreate the patch in cudacpp (as I did for so many other parts of export_cpp that I overload there). I will modify madgraph5/madgraph4gpu#877<https://github.com/madgraph5/madgraph4gpu/pull/877> accordingly. Then the backport from cudacpp to mg5amcnlo can be done when you prefer.

—
Reply to this email directly, view it on GitHub<https://github.com/mg5amcnlo/mg5amcnlo/pull/116#issuecomment-2204310233>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AH6535SVNFHJMYD3G7AFQFLZKMC3TAVCNFSM6AAAAABKFU37S2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEMBUGMYTAMRTGM>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

valassi commented:
Thanks Olivier for the comments.

I understand you do not want to merge this now. I renamed it as WIP and changed it to draft. This will wait for whenever you have time to look at it.

Andrea

> The problem is not where you do the patch here, but that I need a careful check of such PR which is highly non trivial and can not be approve like that either part of the plugin or of mg5amc....

PS I changed https://github.com/madgraph5/madgraph4gpu/pull/877 so that it does not include this PR, as you requested.
I did develop a patch to move this fix to cudacpp, but it is not included (it is reverted) in 877. So 877 will not have this patch enabled, neither in <G5AMC nor in cudacpp.

oliviermattelaer commented:
I have tried to understand your fix here (trying to be constructive) and the real issue is that 
mapconfigs is a list on the fortran side (looks like always)
but that this is not (always?) a list on export_cpp side...

The second weird point is that the cpp side is called before the fortran one which is also a bit weird (and potentially related issue).

While calling cpp or fortran first, is kind of irrelevant, ensure that both cpp and f77 use the exact data structure is to my point of view crucial to avoid side effect and really fix the above issue in a fully save way.

So I'm trying to understand that point but as a preliminary statement, I guess that I would prefer a deeper change to the code to fix such issue (that I volunteer to do).

Cheers,

Olivier

valassi commented:
> So I'm trying to understand that point but as a preliminary statement, I guess that I would prefer a deeper change to the code to fix such issue (that I volunteer to do).

Hi Olivier, thanks a lot fo rlooking at this :-)

As mentioned above, I removed this from https://github.com/madgraph5/madgraph4gpu/pull/877, which is now merged. So fixing this #116 is now not as urgent. 

Instead I am starting to look at the gpucpp_june24 and master_june24 branches, since I understand from you that this is your priority. (I will probably have some questions/observations).

valassi commented:
I am closing this as obsolete.

This is replaced by Olivier's https://github.com/madgraph5/madgraph4gpu/pull/880 which provides a much better fox for https://github.com/madgraph5/madgraph4gpu/issues/856

Thanks Olivier
